plutus-core/plutus-core/src/PlutusCore/Builtin/Convert.hs

```diff
@@ -0,0 +1,533 @@
+              _ -> goLENoLimit (acc <> Builder.storable digitGroup) newRemaining
+    finishLENoLimit :: Builder -> Word64 -> Builder
+    finishLENoLimit acc remaining
+      | remaining == 0 = padLE acc
+      | otherwise = let newRemaining = remaining `unsafeShiftR` 8
+                        digit = fromIntegral remaining
+                      in finishLENoLimit (acc <> Builder.word8 digit) newRemaining
+    padLE :: Builder -> Builder
+    padLE acc = let paddingLength = requestedLength - Builder.builderLength acc
+      in acc <> Builder.bytes (BS.replicate paddingLength 0x0)
+    -- We manually specialize the big-endian case: see Note [Manual specialization] for why.
+    goBELimit :: Builder -> Integer -> Maybe Builder
+    goBELimit acc remaining
+      | remaining == 0 = pure $ padBE acc
+      | otherwise = do
+          guard (Builder.builderLength acc < requestedLength)
+          let newRemaining = remaining `unsafeShiftR` 64
+          let digitGroup = fromInteger remaining
+          case newRemaining of
+            0 -> finishBELimit acc digitGroup
+            _ -> goBELimit (Builder.word64BE digitGroup <> acc) newRemaining
+    finishBELimit :: Builder -> Word64 -> Maybe Builder
+    finishBELimit acc remaining
+      | remaining == 0 = pure $ padBE acc
+      | otherwise = do
+          guard (Builder.builderLength acc < requestedLength)
+          let newRemaining = remaining `unsafeShiftR` 8
+          let digit = fromIntegral remaining
+          finishBELimit (Builder.word8 digit <> acc) newRemaining
+    goBENoLimit :: Builder -> Integer -> Builder
+    goBENoLimit acc remaining
+      | remaining == 0 = padBE acc
+      | otherwise = let newRemaining = remaining `unsafeShiftR` 64
+                        digitGroup = fromInteger remaining
+                      in case newRemaining of
+                        0 -> finishBENoLimit acc digitGroup
+                        _ -> goBENoLimit (Builder.word64BE digitGroup <> acc) newRemaining
+    finishBENoLimit :: Builder -> Word64 -> Builder
+    finishBENoLimit acc remaining
+      | remaining == 0 = padBE acc
+      | otherwise = let newRemaining = remaining `unsafeShiftR` 8
+                        digit = fromIntegral remaining
+                      in finishBENoLimit (Builder.word8 digit <> acc) newRemaining
+    padBE :: Builder -> Builder
+    padBE acc = let paddingLength = requestedLength - Builder.builderLength acc in
+      Builder.bytes (BS.replicate paddingLength 0x0) <> acc
+
+-- | Conversion from 'ByteString' to 'Integer', as per
+-- [CIP-0087](https://github.com/cardano-foundation/CIPs/pull/624).
+--
+-- For clarity, the stated endianness argument uses 'ByteOrder'. Since we only have one failure
+-- condition, we work in 'Maybe', and let the wrapper handle the only possible error.
+byteStringToInteger :: ByteOrder -> ByteString -> Maybe Integer
+byteStringToInteger statedByteOrder bs = do
+  guard (not . BS.null $ bs)
```

Comment:
I spend some time [in the CIP](https://github.com/mlabs-haskell/CIPs/tree/koz/to-from-bytestring/CIP-XXXX#alternative-possibilities) (point 3) explaining why I chose this approach.

Reply:
> I spend some time [in the CIP](https://github.com/mlabs-haskell/CIPs/tree/koz/to-from-bytestring/CIP-XXXX#alternative-possibilities) (point 3) explaining why I chose this approach.

That says
> Typically, the result is suggested to be 0 in such a case. This has some justification: if you view the builtinByteStringToInteger procedure as a fold-with-accumulation over the digits of the representation, this uses the additive monoidal structure of BuiltinInteger, which has 0 as a neutral element.
...
> The fact we could perform the 'reconstruction' with a fold is an implementation detail, rather than an inherent property.

I disagree with that quite strongly. We have to give some semantics for interpreting a bytestring as an integer, and if you're interpreting a bytestring b_0...b_n as an integer (in big-endian form) then the natural interpretation (and exactly the  one used in the CIP) is as \Sigma_{i=0}^n b_i * 256^(n-i).  This isn't an implementation detail, it's the very definition of the conversion function.  In the case of the empty bytestring you get an empty sum, and in mathematics an empty sum is universally agreed to be zero.  I don't see any downside to allowing this.  

However, I'll leave it at that: there's a risk of this just turning into endless bickering about an edge case and I don't think that that would be very productive.

Reply:
I find @kwxm's argument persuasive, and have since amended both the CIP and the implementation. It also enables better properties, so I feel it's a win overall.